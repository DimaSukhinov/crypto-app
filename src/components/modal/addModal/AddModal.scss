.modal {
  display: flex;
  flex-direction: column;
  align-items: center;
  min-width: 300px;

  &__item {
    margin: 10px;
  }
}
