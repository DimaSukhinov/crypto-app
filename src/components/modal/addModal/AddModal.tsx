import React, {ChangeEvent, useCallback} from 'react';
import './AddModal.scss';
import {Modal} from '../Modal';
import {ValueType} from '../../../store/values-reducer';
import {addToPortfolioAC, PortfolioType} from '../../../store/portfolio-reducer';
import {useDispatch} from 'react-redux';

type AddModalPropsType = {
    values: ValueType[]
    activeAddModal: boolean
    currentValue: string
    valueCount: number
    error: boolean
    setError: (error: boolean) => void
    setValueCount: (valueCount: number) => void
    setActiveAddModal: (activeAddModal: boolean) => void
    onValueCountChange: (e: ChangeEvent<HTMLInputElement>) => void
}

export const AddModal = React.memo(({
                                        values, activeAddModal, setActiveAddModal, error, onValueCountChange,
                                        setValueCount, currentValue, setError, valueCount
                                    }: AddModalPropsType) => {

    const dispatch = useDispatch()

    const addToPortfolio = useCallback((id: string, name: string, price: string, valueCount: number) => () => {
        if (valueCount > 0) {
            let value: PortfolioType = {id: id, name: name, price: +price, valueCount: valueCount}
            dispatch(addToPortfolioAC(value))
            setActiveAddModal(false)
            setValueCount(0)
        } else setError(true)
    }, [dispatch, setActiveAddModal, setError, setValueCount])

    return (
        <>
            {activeAddModal && <Modal setActive={setActiveAddModal}>
                {values.map(v => v.id === currentValue && <div className={'modal'}>
                    <span className={'modal__item'}>{v.name}</span>
                    <input type="number" onChange={onValueCountChange} className={'modal__item'}/>
                    {error && <div style={{color: 'red'}}>Incorrect value</div>}
                    <span className={'modal__item'}>
                         Price: {valueCount > 0 && (valueCount * +v.priceUsd).toFixed(2) + '$'}
                    </span>
                    <div onClick={addToPortfolio(v.id, v.name, v.priceUsd, valueCount)}
                         className={'modal__item values__add-button'}>Add
                    </div>
                </div>)}
            </Modal>}
        </>
    );
})
