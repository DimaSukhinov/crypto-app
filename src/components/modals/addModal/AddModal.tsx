import React, { ChangeEvent } from 'react';
import './AddModal.scss';
import { Modal } from '../Modal';
import { ValueType } from '../../../store/values-reducer';
import { Button } from '../../common/button/Button';

type AddModalPropsType = {
  values: ValueType[]
  activeAddModal: boolean
  currentValue: string
  closeModal: () => void
  error: boolean
  valueCount: number
  onValueCountChange: (e: ChangeEvent<HTMLInputElement>) => void
  addToPortfolio: (id: string, name: string, price: string, valueCount: number) => () => void
}

export const AddModal = React.memo(({
                                      values, activeAddModal, closeModal, currentValue,
                                      valueCount, error, onValueCountChange, addToPortfolio,
                                    }: AddModalPropsType) => {

  return (
    <div data-testid={'add-modal'}>
      {activeAddModal && <Modal closeModal={closeModal}>
        {values.map(v => v.id === currentValue && <div key={v.id} className={'modal'}>
          <span className={'modal__item'}>{v.name}</span>
          <input type='number' onChange={onValueCountChange} className={'modal__item'} />
          {error && <div style={{ color: 'red' }} className={'modal__error'}>min 0.1, max 1000</div>}
          <span className={'modal__item'}>
            Price: {valueCount > 0 && (valueCount * +v.priceUsd).toFixed(2) + '$'}
          </span>
          <Button onClickHandler={addToPortfolio(v.id, v.name, v.priceUsd, valueCount)}>
            <span className={'modal__buttonText'}>Add</span>
          </Button>
        </div>)}
      </Modal>}
    </div>
  );
});
