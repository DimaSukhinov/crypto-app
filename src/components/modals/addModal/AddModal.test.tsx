import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, screen } from '@testing-library/react';
import { AddModal } from './AddModal';

describe('Add modal', () => {

  const value = {
    id: 'bitcoin', rank: '1', changePercent24Hr: '3.1367184772434482', explorer: 'https://blockchain.info/',
    marketCapUsd: '452445368357.8622563021714334', maxSupply: '21000000.0000000000000000', name: 'Bitcoin',
    priceUsd: '23679.9797033025578357', supply: '19106662.0000000000000000', symbol: 'BTC',
    volumeUsd24Hr: '17521511686.5773940471774064', vwap24Hr: '23710.9068155355389843',
  };
  const currentValue = 'bitcoin';

  test('without error', () => {
    render(
      <AddModal closeModal={jest.fn()} values={[value]} activeAddModal={true} valueCount={0} error={false}
                addToPortfolio={jest.fn()} onValueCountChange={jest.fn()} currentValue={currentValue}
      />,
    );

    expect(screen.getByTestId('add-modal')).toBeInTheDocument();
    expect(screen.getByTestId('add-modal')).toMatchSnapshot();

    expect(screen.getByText(/price/i)).toBeInTheDocument();
    expect(screen.getByText(/add/i)).toBeInTheDocument();
    expect(screen.queryByText(/incorrect value/i)).not.toBeInTheDocument();
  });

  test('with error', () => {
    render(
      <AddModal closeModal={jest.fn()} values={[value]} activeAddModal={true} currentValue={currentValue}
                addToPortfolio={jest.fn()} error={true} onValueCountChange={jest.fn()} valueCount={0}
      />,
    );

    expect(screen.getByTestId('add-modal')).toBeInTheDocument();
    expect(screen.getByTestId('add-modal')).toMatchSnapshot();

    expect(screen.getByText(/min 0.1, max 1000/i)).toBeInTheDocument();
    expect(screen.getByText(/min 0.1, max 1000/i)).toHaveStyle({ color: 'red' });
  });
});
