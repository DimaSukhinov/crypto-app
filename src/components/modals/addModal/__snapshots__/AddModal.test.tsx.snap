// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add modal with error 1`] = `
<div
  data-testid="add-modal"
>
  <div
    class="reusableModal"
    data-testid="space-around-modal"
  >
    <div
      class="reusableModal__content"
    >
      <div
        class="modal"
      >
        <span
          class="modal__item"
        >
          Bitcoin
        </span>
        <input
          class="modal__item"
          type="number"
        />
        <div
          class="modal__error"
          style="color: red;"
        >
          min 0.1, max 1000
        </div>
        <span
          class="modal__item"
        >
          Price: 
        </span>
        <div
          class="button"
        >
          <span
            class="modal__buttonText"
          >
            Add
          </span>
        </div>
      </div>
      <div
        class="reusableModal__close"
      >
        X
      </div>
    </div>
  </div>
</div>
`;

exports[`Add modal without error 1`] = `
<div
  data-testid="add-modal"
>
  <div
    class="reusableModal"
    data-testid="space-around-modal"
  >
    <div
      class="reusableModal__content"
    >
      <div
        class="modal"
      >
        <span
          class="modal__item"
        >
          Bitcoin
        </span>
        <input
          class="modal__item"
          type="number"
        />
        <span
          class="modal__item"
        >
          Price: 
        </span>
        <div
          class="button"
        >
          <span
            class="modal__buttonText"
          >
            Add
          </span>
        </div>
      </div>
      <div
        class="reusableModal__close"
      >
        X
      </div>
    </div>
  </div>
</div>
`;
